### ISSUE TITLE
Cannot read property 'layerContainer' of undefined

### ISSUE THREAD

**Issue by bslipek** [mhasbie/react-leaflet-deflate#2]
````
"leaflet": "^1.3.4",
"react-leaflet": "^1.9.1",
"react-leaflet-deflate": "^2.1.0",
````

I get `Cannot read property 'layerContainer' of undefined`.

![react_app](https://user-images.githubusercontent.com/2363946/46954398-c0019f80-d090-11e8-9d9f-a20c2f50c8f3.png)

I belive here is the problem. Shouldn't`layerContainer` come from `context`?

**Comment by Sobak** [comment]
:+1:

**Comment by bslipek** [comment]
I make it works like this:
```
import Leaflet from 'leaflet';
import DeflatedFeatureGroup from 'Leaflet.Deflate';
import { MapLayer } from 'react-leaflet';
import 'leaflet.markercluster';


export default class Deflate extends MapLayer {

	createLeafletElement(props) {
    return Leaflet.deflate(props);
	}

	componentDidMount() {
		const { map } = this.props.leaflet || this.context;
		const { markerCluster } = this.props;
		// This will add our DeflatedFeatureGroup (this.leafletElement) to the map--something that must happen BEFORE
		// we can add anything to DeflatedFeatureGroup itself.
		super.componentDidMount();
		Leaflet.geoJSON(this.props.data, this.props).addTo(this.leafletElement);
		if (markerCluster) map._container.className += ' leaflet-cluster-anim marker-cluster-styled marker-cluster-animated';
	}

	updateLeafletElement(fromProps, toProps) {
		this.context.layerContainer.removeLayer(this.leafletElement);
		this.leafletElement = Leaflet.deflate(toProps);
		this.context.layerContainer.addLayer(this.leafletElement);
		Leaflet.geoJSON(toProps.data, toProps).addTo(this.leafletElement);
	}
}
```

But now there is a problem when applying `removeLayer` -> `	obj._leaflet_id = obj._leaflet_id || ++lastId;`

**Comment by mhasbie** [comment]
Hi,

Can you post your codes here or a jsfiddle or codepen to replicate this?

Thanks

**Comment by bslipek** [comment]
@mhasbie I will try :) thanks

**Comment by bslipek** [comment]
@mhasbie it is very hard to reproduce  the error, because we need to install and then import libs.

try go here: https://codesandbox.io/s/210ov3oopj?expanddevtools=1

and the after page is loaded try make a change in the code editor, for example make press space there. You should see this:

![210ov3oopj_-_codesandbox](https://user-images.githubusercontent.com/2363946/47089813-e1de5c00-d221-11e8-8816-5fa95e5247ca.png)

This is the error what I now have in my code also ;/

**Comment by bslipek** [comment]
Here is it!

[![Edit 5vr1yyyxkx](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/5vr1yyyxkx)

Can you make this async example works? (Just wait 5sec for crash)

**Comment by daszkowski** [comment]
+1

**Comment by slaby93** [comment]
+1

**Comment by bslipek** [comment]
Here is some progress:
1. I can add features dynamically
2. There should be markers but can;t display it in codesanbox, works locally ;/

[![Edit 5yq6pv2nvp](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/5yq6pv2nvp)

But:
**1. still can't make `markerCluster={true}`**
**2. when you place space somewhere in the editor error appear, probably related to rerender** component.

**Comment by bslipek** [comment]
Here is fixed version with new way of remove layers, not perfect but works:

```js
removeLayers() {
    let layers = this.layerContainer._layers;
    let layer_ids = Object.keys(layers);

    layer_ids.forEach(id => {
      let layer = layers[id];
      if (layer._url) return; // <- prevent removing tile layer, only features group, not nice ;/
      this.layerContainer.removeLayer(layer);
    });
  }
```

[![Edit React Leaflet Deflate Fixed](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/6x7o943wkz)

We still can't use `markerCluster`. To do this we need to modify `Leaflet` library:
```js
// leaflet-src.js line 70
function stamp(obj) {
  if (!obj) obj = {}; // <- here
  obj._leaflet_id = obj._leaflet_id || ++lastId;

  return obj._leaflet_id;
}
```

**Comment by bslipek** [comment]
I created PR for Leaflet lib.
https://github.com/Leaflet/Leaflet/pull/6381

**Comment by mhasbie** [comment]
This issue is now fixed in v2.2.0. Thanks for reporting.

**Comment by bslipek** [comment]
It is much better now!
One thing I want to suggest: 
````js
// Deflate.js line 53
if (toProps.data && toProps.data.features && toProps.data.features.length > 0) 
   Leaflet.geoJSON(toProps.data, toProps).addTo(this.leafletElement);
````
because when we try add empty FeatureCollection then Leaflet.Deflate thow error.
